workerpool_test.go

```
@@ -402,3 +402,60 @@ func TestWorkerPoolClose(t *testing.T) {
 	}
 	wg.Wait() // all routines should have returned
 }
+
+func TestWorkerPoolNewWithContext(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	n := runtime.NumCPU()
+	wp := workerpool.NewWithContext(ctx, n)
+
+	// working is written to by each task as soon as possible.
+	working := make(chan struct{})
+	var wg sync.WaitGroup
+	// Create n tasks waiting on the context to be cancelled.
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		id := fmt.Sprintf("task #%2d", i)
+		err := wp.Submit(id, func(ctx context.Context) error {
+			working <- struct{}{}
+			<-ctx.Done()
+			wg.Done()
+			return ctx.Err()
+		})
+		if err != nil {
+			t.Errorf("failed to submit task '%s': %v", id, err)
+		}
+	}
+
+	// ensure n workers are busy
+	for i := 0; i < n; i++ {
+		<-working
+	}
+
+	// cancel the parent context, which should complete all tasks.
+	cancel()
+	wg.Wait()
+
+	// Submitting a task once the parent context has been cancelled should
+	// succeed and give a cancelled context to the task. This is not ideal and
+	// might change in the future.
```

What is the problem with this? Why is it not ideal?

---

IMO ideally `Submit` should return an error when the parent context is cancelled, because what would do the task do when given an already cancelled context?

To do that, `Submit` would have to check the context status but I didn't feel like keeping a context reference into the workerpool struct in this PR.

---

I feel bad for the task to be given a cancelled context before it could do anything, but ok.

---

Yes, I agree. If the caller stop submitting tasks as soon as the context is cancelled, then what happen on Submit doesn't matter. That's why I think this PR implementation is an acceptable behavior.

But if — for whatever reason — the caller still Submit when the context is cancelled, it would be ideal to shortcut and return an error. I think debugging would be easier if it was the case.